// appium-doctor/build/lib/android.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _doctor = require('./doctor');

var _utils = require('./utils');

var _appiumSupport = require('appium-support');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var checks = [];

var javaHome = _appiumSupport.system.isWindows() ? '%JAVA_HOME%' : '$JAVA_HOME';

checks.push(new _env2['default']('ANDROID_HOME'));
checks.push(new _env2['default']('JAVA_HOME'));

// Check that the PATH includes the jdk's bin directory

var JavaOnPathCheck = (function (_DoctorCheck) {
  _inherits(JavaOnPathCheck, _DoctorCheck);

  function JavaOnPathCheck() {
    _classCallCheck(this, JavaOnPathCheck);

    _get(Object.getPrototypeOf(JavaOnPathCheck.prototype), 'constructor', this).apply(this, arguments);
  }

  // Check tools

  _createClass(JavaOnPathCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var javaHomeBin;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!process.env.JAVA_HOME) {
              context$2$0.next = 4;
              break;
            }

            javaHomeBin = _path2['default'].resolve(process.env.JAVA_HOME, 'bin');

            if (!(process.env.PATH.indexOf(javaHomeBin) + 1)) {
              context$2$0.next = 4;
              break;
            }

            return context$2$0.abrupt('return', (0, _utils.ok)('Bin directory of ' + javaHome + ' is set'));

          case 4:
            return context$2$0.abrupt('return', (0, _utils.nok)('Bin directory for ' + javaHome + ' is not set'));

          case 5:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return 'Add \'' + javaHome + _path2['default'].sep + 'bin\' to your PATH environment';
    }
  }]);

  return JavaOnPathCheck;
})(_doctor.DoctorCheck);

var AndroidToolCheck = (function (_DoctorCheck2) {
  _inherits(AndroidToolCheck, _DoctorCheck2);

  function AndroidToolCheck(toolName, toolPath) {
    _classCallCheck(this, AndroidToolCheck);

    _get(Object.getPrototypeOf(AndroidToolCheck.prototype), 'constructor', this).call(this);
    this.toolName = toolName;
    this.toolPath = toolPath;
  }

  _createClass(AndroidToolCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var fullPath;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(typeof process.env.ANDROID_HOME === 'undefined')) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return', (0, _utils.nok)(this.toolName + ' could not be found because ANDROID_HOME is NOT set!'));

          case 2:
            fullPath = _path2['default'].resolve(process.env.ANDROID_HOME, this.toolPath);
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(fullPath));

          case 5:
            if (!context$2$0.sent) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.t0 = (0, _utils.ok)(this.toolName + ' exists at: ' + fullPath);
            context$2$0.next = 10;
            break;

          case 9:
            context$2$0.t0 = (0, _utils.nok)(this.toolName + ' could NOT be found at \'' + fullPath + '\'!');

          case 10:
            return context$2$0.abrupt('return', context$2$0.t0);

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      if (typeof process.env.ANDROID_HOME === 'undefined') {
        return 'Manually configure ANDROID_HOME and run appium-doctor again.';
      }
      return 'Manually install ' + this.toolName + ' and add it to PATH.';
    }
  }]);

  return AndroidToolCheck;
})(_doctor.DoctorCheck);

checks.push(new AndroidToolCheck('adb', _path2['default'].join("platform-tools", _appiumSupport.system.isWindows() ? 'adb.exe' : 'adb')));
checks.push(new AndroidToolCheck('android', _path2['default'].join("tools", _appiumSupport.system.isWindows() ? 'android.bat' : 'android')));
checks.push(new AndroidToolCheck('emulator', _path2['default'].join("tools", _appiumSupport.system.isWindows() ? 'emulator.exe' : 'emulator')));
checks.push(new JavaOnPathCheck());

exports.EnvVarAndPathCheck = _env2['default'];
exports.AndroidToolCheck = AndroidToolCheck;
exports.JavaOnPathCheck = JavaOnPathCheck;
exports['default'] = checks;
