// appium-doctor/build/lib/demo.js
// demo rule to test the gui

'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _utils = require('./utils');

var _appiumSupport = require('appium-support');

var _teen_process = require('teen_process');

var _doctor = require('./doctor');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _prompt = require('./prompt');

var checks = [];

var DirCheck = (function (_DoctorCheck) {
  _inherits(DirCheck, _DoctorCheck);

  function DirCheck(path) {
    _classCallCheck(this, DirCheck);

    _get(Object.getPrototypeOf(DirCheck.prototype), 'constructor', this).call(this, { autofix: false });
    this.path = path;
  }

  _createClass(DirCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var stats;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(this.path));

          case 2:
            if (context$2$0.sent) {
              context$2$0.next = 4;
              break;
            }

            return context$2$0.abrupt('return', (0, _utils.nok)('Could NOT find directory at \'' + this.path + '\'!'));

          case 4:
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.lstat(this.path));

          case 6:
            stats = context$2$0.sent;
            return context$2$0.abrupt('return', stats.isDirectory() ? (0, _utils.ok)('Found directory at: ' + this.path) : (0, _utils.nok)('\'' + this.path + '\' is NOT a directory!'));

          case 8:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', 'Manually create a directory at: ' + this.path);

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return DirCheck;
})(_doctor.DoctorCheck);

checks.push(new DirCheck('/tmp/appium-doctor'));
checks.push(new DirCheck('/tmp/appium-doctor/demo'));

var FileCheck = (function (_DoctorCheck2) {
  _inherits(FileCheck, _DoctorCheck2);

  function FileCheck(path) {
    _classCallCheck(this, FileCheck);

    _get(Object.getPrototypeOf(FileCheck.prototype), 'constructor', this).call(this, { autofix: true });
    this.path = path;
  }

  _createClass(FileCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(this.path));

          case 2:
            if (!context$2$0.sent) {
              context$2$0.next = 6;
              break;
            }

            context$2$0.t0 = (0, _utils.ok)('Found file at: ' + this.path);
            context$2$0.next = 7;
            break;

          case 6:
            context$2$0.t0 = (0, _utils.nok)('Could NOT find file at \'' + this.path + '\'!');

          case 7:
            return context$2$0.abrupt('return', context$2$0.t0);

          case 8:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      var yesno;
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].info('The following command need be executed: touch \'' + this.path + '\'');
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap((0, _prompt.fixIt)());

          case 3:
            yesno = context$2$0.sent;

            if (!(yesno === 'yes')) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.next = 7;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('touch', [this.path]));

          case 7:
            context$2$0.next = 11;
            break;

          case 9:
            _logger2['default'].info('Skipping you will need to touch \'' + this.path + '\' manually.');
            throw new _doctor.FixSkippedError('bbb');

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return FileCheck;
})(_doctor.DoctorCheck);

checks.push(new FileCheck('/tmp/appium-doctor/demo/apple.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/pear.fruit'));
checks.push(new FileCheck('/tmp/appium-doctor/demo/orange.fruit'));

exports.DirCheck = DirCheck;
exports.FileCheck = FileCheck;
exports['default'] = checks;
