// appium-doctor/build/lib/dev.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _teen_process = require('teen_process');

var _doctor = require('./doctor');

var _utils = require('./utils');

var _appiumSupport = require('appium-support');

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _os = require('os');

var checks = [];

// Check PATH binaries

var BinaryIsInPathCheck = (function (_DoctorCheck) {
  _inherits(BinaryIsInPathCheck, _DoctorCheck);

  function BinaryIsInPathCheck(binary) {
    _classCallCheck(this, BinaryIsInPathCheck);

    _get(Object.getPrototypeOf(BinaryIsInPathCheck.prototype), 'constructor', this).call(this);
    this.binary = binary;
  }

  _createClass(BinaryIsInPathCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var resolvedPath, executable, _ref, stdout;

      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            resolvedPath = undefined;
            context$2$0.prev = 1;
            executable = _appiumSupport.system.isWindows() ? 'where' : 'which';
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)(executable, [this.binary]));

          case 5:
            _ref = context$2$0.sent;
            stdout = _ref.stdout;

            if (!stdout.match(/not found/gi)) {
              context$2$0.next = 9;
              break;
            }

            throw new Error('Not Found');

          case 9:
            resolvedPath = _appiumSupport.system.isWindows() ? stdout.split(_os.EOL)[0] : stdout.replace(_os.EOL, '');
            context$2$0.next = 15;
            break;

          case 12:
            context$2$0.prev = 12;
            context$2$0.t0 = context$2$0['catch'](1);
            return context$2$0.abrupt('return', (0, _utils.nok)(this.binary + ' is MISSING in PATH!'));

          case 15:
            context$2$0.next = 17;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(resolvedPath));

          case 17:
            if (!context$2$0.sent) {
              context$2$0.next = 21;
              break;
            }

            context$2$0.t1 = (0, _utils.ok)(this.binary + ' was found at ' + resolvedPath);
            context$2$0.next = 22;
            break;

          case 21:
            context$2$0.t1 = (0, _utils.nok)(this.binary + ' was found in PATH at \'' + resolvedPath + '\', but this is NOT a valid path!');

          case 22:
            return context$2$0.abrupt('return', context$2$0.t1);

          case 23:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 12]]);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return 'Manually install the ' + this.binary + ' binary and add it to PATH.';
    }
  }]);

  return BinaryIsInPathCheck;
})(_doctor.DoctorCheck);

checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'mvn.bat' : 'mvn'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'ant.bat' : 'ant'));
checks.push(new BinaryIsInPathCheck(_appiumSupport.system.isWindows() ? 'adb.exe' : 'adb'));

// Check Android SDKs

var AndroidSdkExists = (function (_DoctorCheck2) {
  _inherits(AndroidSdkExists, _DoctorCheck2);

  function AndroidSdkExists(sdk) {
    _classCallCheck(this, AndroidSdkExists);

    _get(Object.getPrototypeOf(AndroidSdkExists.prototype), 'constructor', this).call(this);
    this.sdk = sdk;
  }

  _createClass(AndroidSdkExists, [{
    key: 'diagnose',
    value: function diagnose() {
      var sdkPath;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(typeof process.env.ANDROID_HOME === 'undefined')) {
              context$2$0.next = 2;
              break;
            }

            return context$2$0.abrupt('return', (0, _utils.nok)(this.sdk + ' could not be found because ANDROID_HOME is NOT set!'));

          case 2:
            sdkPath = _path2['default'].resolve(process.env.ANDROID_HOME, _path2['default'].join("platforms", this.sdk));
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(sdkPath));

          case 5:
            if (!context$2$0.sent) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.t0 = (0, _utils.ok)(this.sdk + ' was found at: ' + sdkPath);
            context$2$0.next = 10;
            break;

          case 9:
            context$2$0.t0 = (0, _utils.nok)(this.sdk + ' could NOT be found at \'' + sdkPath + '\'!');

          case 10:
            return context$2$0.abrupt('return', context$2$0.t0);

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      if (typeof process.env.ANDROID_HOME === "undefined") {
        return 'Manually configure ANDROID_HOME.';
      }
      return 'Manually install the ' + this.sdk + ' sdk.';
    }
  }]);

  return AndroidSdkExists;
})(_doctor.DoctorCheck);

checks.push(new AndroidSdkExists('android-16'));
checks.push(new AndroidSdkExists('android-19'));

exports.BinaryIsInPathCheck = BinaryIsInPathCheck;
exports.AndroidSdkExists = AndroidSdkExists;
exports['default'] = checks;
