// appium-doctor/build/lib/doctor.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

require('colors');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _packageJson = require('../../package.json');

// eslint-disable-line import/no-unresolved

var FixSkippedError = (function (_Error) {
  _inherits(FixSkippedError, _Error);

  function FixSkippedError() {
    _classCallCheck(this, FixSkippedError);

    _get(Object.getPrototypeOf(FixSkippedError.prototype), 'constructor', this).apply(this, arguments);
  }

  return FixSkippedError;
})(Error);

var DoctorCheck = (function () {
  function DoctorCheck() {
    var opts = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

    _classCallCheck(this, DoctorCheck);

    this.autofix = !!opts.autofix;
  }

  _createClass(DoctorCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      throw new Error('Not Implemented!');
    }
  }, {
    key: 'fix',
    value: function fix() {
      // return string for manual fixes.
      throw new Error('Not Implemented!');
    }
  }]);

  return DoctorCheck;
})();

var Doctor = (function () {
  function Doctor() {
    _classCallCheck(this, Doctor);

    this.checks = [];
    this.toFix = [];
  }

  _createClass(Doctor, [{
    key: 'register',
    value: function register(checks) {
      checks = Array.isArray(checks) ? checks : [checks];
      this.checks = this.checks.concat(checks);
    }
  }, {
    key: 'diagnose',
    value: function diagnose() {
      var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, check, res, errorMessage, fixMessage;

      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].info('### Diagnostic starting ###');
            this.toFix = [];
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            context$2$0.prev = 5;
            _iterator = _getIterator(this.checks);

          case 7:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              context$2$0.next = 16;
              break;
            }

            check = _step.value;
            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(check.diagnose());

          case 11:
            res = context$2$0.sent;

            if (res.ok) {
              _logger2['default'].info(' ' + '✔'.green + ' ' + res.message);
            } else {
              errorMessage = ' ' + '✖'.red + ' ' + res.message;

              _logger2['default'].warn(errorMessage);
              this.toFix.push({
                error: errorMessage,
                check: check
              });
            }

          case 13:
            _iteratorNormalCompletion = true;
            context$2$0.next = 7;
            break;

          case 16:
            context$2$0.next = 22;
            break;

          case 18:
            context$2$0.prev = 18;
            context$2$0.t0 = context$2$0['catch'](5);
            _didIteratorError = true;
            _iteratorError = context$2$0.t0;

          case 22:
            context$2$0.prev = 22;
            context$2$0.prev = 23;

            if (!_iteratorNormalCompletion && _iterator['return']) {
              _iterator['return']();
            }

          case 25:
            context$2$0.prev = 25;

            if (!_didIteratorError) {
              context$2$0.next = 28;
              break;
            }

            throw _iteratorError;

          case 28:
            return context$2$0.finish(25);

          case 29:
            return context$2$0.finish(22);

          case 30:
            fixMessage = undefined;
            context$2$0.t1 = this.toFix.length;
            context$2$0.next = context$2$0.t1 === 0 ? 34 : context$2$0.t1 === 1 ? 36 : 38;
            break;

          case 34:
            fixMessage = 'no fix needed';
            return context$2$0.abrupt('break', 39);

          case 36:
            fixMessage = 'one fix needed';
            return context$2$0.abrupt('break', 39);

          case 38:
            fixMessage = this.toFix.length + ' fixes needed';

          case 39:
            _logger2['default'].info('### Diagnostic completed, ' + fixMessage + '. ###');
            _logger2['default'].info('');

          case 41:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[5, 18, 22, 30], [23,, 25, 29]]);
    }
  }, {
    key: 'reportSuccess',
    value: function reportSuccess() {
      return _regeneratorRuntime.async(function reportSuccess$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (!(this.toFix.length === 0)) {
              context$2$0.next = 6;
              break;
            }

            _logger2['default'].info('Everything looks good, bye!');
            _logger2['default'].info('');
            return context$2$0.abrupt('return', true);

          case 6:
            return context$2$0.abrupt('return', false);

          case 7:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'reportManualFixes',
    value: function reportManualFixes() {
      var manualFixes, fixMessages, _iteratorNormalCompletion2, _didIteratorError2, _iteratorError2, _iterator2, _step2, f, _iteratorNormalCompletion3, _didIteratorError3, _iteratorError3, _iterator3, _step3, m;

      return _regeneratorRuntime.async(function reportManualFixes$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            manualFixes = _lodash2['default'].filter(this.toFix, function (f) {
              return !f.check.autofix;
            });

            if (!(manualFixes.length > 0)) {
              context$2$0.next = 60;
              break;
            }

            _logger2['default'].info('### Manual Fixes Needed ###');
            _logger2['default'].info('The configuration cannot be automatically fixed, please do the following first:');
            // for manual fixes, the fix method always return a string
            fixMessages = [];
            _iteratorNormalCompletion2 = true;
            _didIteratorError2 = false;
            _iteratorError2 = undefined;
            context$2$0.prev = 8;
            _iterator2 = _getIterator(manualFixes);

          case 10:
            if (_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done) {
              context$2$0.next = 20;
              break;
            }

            f = _step2.value;
            context$2$0.t0 = fixMessages;
            context$2$0.next = 15;
            return _regeneratorRuntime.awrap(f.check.fix());

          case 15:
            context$2$0.t1 = context$2$0.sent;
            context$2$0.t0.push.call(context$2$0.t0, context$2$0.t1);

          case 17:
            _iteratorNormalCompletion2 = true;
            context$2$0.next = 10;
            break;

          case 20:
            context$2$0.next = 26;
            break;

          case 22:
            context$2$0.prev = 22;
            context$2$0.t2 = context$2$0['catch'](8);
            _didIteratorError2 = true;
            _iteratorError2 = context$2$0.t2;

          case 26:
            context$2$0.prev = 26;
            context$2$0.prev = 27;

            if (!_iteratorNormalCompletion2 && _iterator2['return']) {
              _iterator2['return']();
            }

          case 29:
            context$2$0.prev = 29;

            if (!_didIteratorError2) {
              context$2$0.next = 32;
              break;
            }

            throw _iteratorError2;

          case 32:
            return context$2$0.finish(29);

          case 33:
            return context$2$0.finish(26);

          case 34:
            _iteratorNormalCompletion3 = true;
            _didIteratorError3 = false;
            _iteratorError3 = undefined;
            context$2$0.prev = 37;

            for (_iterator3 = _getIterator(_lodash2['default'].uniq(fixMessages)); !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
              m = _step3.value;

              _logger2['default'].warn('- ' + m);
            }
            context$2$0.next = 45;
            break;

          case 41:
            context$2$0.prev = 41;
            context$2$0.t3 = context$2$0['catch'](37);
            _didIteratorError3 = true;
            _iteratorError3 = context$2$0.t3;

          case 45:
            context$2$0.prev = 45;
            context$2$0.prev = 46;

            if (!_iteratorNormalCompletion3 && _iterator3['return']) {
              _iterator3['return']();
            }

          case 48:
            context$2$0.prev = 48;

            if (!_didIteratorError3) {
              context$2$0.next = 51;
              break;
            }

            throw _iteratorError3;

          case 51:
            return context$2$0.finish(48);

          case 52:
            return context$2$0.finish(45);

          case 53:
            _logger2['default'].info('###');
            _logger2['default'].info('');
            _logger2['default'].info('Bye! Run appium-doctor again when all manual fixes have been applied!');
            _logger2['default'].info('');
            return context$2$0.abrupt('return', true);

          case 60:
            return context$2$0.abrupt('return', false);

          case 61:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[8, 22, 26, 34], [27,, 29, 33], [37, 41, 45, 53], [46,, 48, 52]]);
    }
  }, {
    key: 'runAutoFix',
    value: function runAutoFix(f) {
      var res;
      return _regeneratorRuntime.async(function runAutoFix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].info('### Fixing: ' + f.error + ' ###');
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap(f.check.fix());

          case 4:
            context$2$0.next = 16;
            break;

          case 6:
            context$2$0.prev = 6;
            context$2$0.t0 = context$2$0['catch'](1);

            if (!(context$2$0.t0 instanceof FixSkippedError)) {
              context$2$0.next = 13;
              break;
            }

            _logger2['default'].info('### Skipped fix ###');
            return context$2$0.abrupt('return');

          case 13:
            _logger2['default'].warn(('' + context$2$0.t0).replace(/\n$/g, ''));
            _logger2['default'].info('### Fix did not succeed ###');
            return context$2$0.abrupt('return');

          case 16:
            _logger2['default'].info('Checking if this was fixed:');
            context$2$0.next = 19;
            return _regeneratorRuntime.awrap(f.check.diagnose());

          case 19:
            res = context$2$0.sent;

            if (res.ok) {
              f.fixed = true;
              _logger2['default'].info(' ' + '✔'.green + ' ' + res.message);
              _logger2['default'].info('### Fix was successfully applied ###');
            } else {
              _logger2['default'].info(' ' + '✖'.red + ' ' + res.message);
              _logger2['default'].info('### Fix was applied but issue remains ###');
            }

          case 21:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 6]]);
    }
  }, {
    key: 'runAutoFixes',
    value: function runAutoFixes() {
      var autoFixes, _iteratorNormalCompletion4, _didIteratorError4, _iteratorError4, _iterator4, _step4, f;

      return _regeneratorRuntime.async(function runAutoFixes$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            autoFixes = _lodash2['default'].filter(this.toFix, function (f) {
              return f.check.autofix;
            });
            _iteratorNormalCompletion4 = true;
            _didIteratorError4 = false;
            _iteratorError4 = undefined;
            context$2$0.prev = 4;
            _iterator4 = _getIterator(autoFixes);

          case 6:
            if (_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done) {
              context$2$0.next = 14;
              break;
            }

            f = _step4.value;
            context$2$0.next = 10;
            return _regeneratorRuntime.awrap(this.runAutoFix(f));

          case 10:
            _logger2['default'].info('');

          case 11:
            _iteratorNormalCompletion4 = true;
            context$2$0.next = 6;
            break;

          case 14:
            context$2$0.next = 20;
            break;

          case 16:
            context$2$0.prev = 16;
            context$2$0.t0 = context$2$0['catch'](4);
            _didIteratorError4 = true;
            _iteratorError4 = context$2$0.t0;

          case 20:
            context$2$0.prev = 20;
            context$2$0.prev = 21;

            if (!_iteratorNormalCompletion4 && _iterator4['return']) {
              _iterator4['return']();
            }

          case 23:
            context$2$0.prev = 23;

            if (!_didIteratorError4) {
              context$2$0.next = 26;
              break;
            }

            throw _iteratorError4;

          case 26:
            return context$2$0.finish(23);

          case 27:
            return context$2$0.finish(20);

          case 28:
            if (_lodash2['default'].find(autoFixes, function (f) {
              return !f.fixed;
            })) {
              // a few issues remain.
              _logger2['default'].info('Bye! A few issues remain, fix manually and/or rerun appium-doctor!');
            } else {
              // nothing left to fix.
              _logger2['default'].info('Bye! All issues have been fixed!');
            }
            _logger2['default'].info('');

          case 30:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[4, 16, 20, 28], [21,, 23, 27]]);
    }
  }, {
    key: 'run',
    value: function run() {
      return _regeneratorRuntime.async(function run$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].info('Appium Doctor v.' + _packageJson.version);
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(this.diagnose());

          case 3:
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(this.reportSuccess());

          case 5:
            if (!context$2$0.sent) {
              context$2$0.next = 7;
              break;
            }

            return context$2$0.abrupt('return');

          case 7:
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(this.reportManualFixes());

          case 9:
            if (!context$2$0.sent) {
              context$2$0.next = 11;
              break;
            }

            return context$2$0.abrupt('return');

          case 11:
            context$2$0.next = 13;
            return _regeneratorRuntime.awrap(this.runAutoFixes());

          case 13:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return Doctor;
})();

exports.Doctor = Doctor;
exports.DoctorCheck = DoctorCheck;
exports.FixSkippedError = FixSkippedError;
