// appium-doctor/build/lib/ios.js
'use strict';

var _get = require('babel-runtime/helpers/get')['default'];

var _inherits = require('babel-runtime/helpers/inherits')['default'];

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _utils = require('./utils');

var _appiumSupport = require('appium-support');

var _teen_process = require('teen_process');

var _doctor = require('./doctor');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _carthageDetector = require('./carthage-detector');

var _carthageDetector2 = _interopRequireDefault(_carthageDetector);

var _prompt = require('./prompt');

var _env = require('./env');

var _env2 = _interopRequireDefault(_env);

var checks = [];
var fixes = {};

// Check for Xcode.

var XcodeCheck = (function (_DoctorCheck) {
  _inherits(XcodeCheck, _DoctorCheck);

  function XcodeCheck() {
    _classCallCheck(this, XcodeCheck);

    _get(Object.getPrototypeOf(XcodeCheck.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(XcodeCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var xcodePath, _ref, stdout;

      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            xcodePath = undefined;
            context$2$0.prev = 1;
            context$2$0.next = 4;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('xcode-select', ['--print-path']));

          case 4:
            _ref = context$2$0.sent;
            stdout = _ref.stdout;

            xcodePath = (stdout || '').replace("\n", "");
            context$2$0.next = 12;
            break;

          case 9:
            context$2$0.prev = 9;
            context$2$0.t0 = context$2$0['catch'](1);
            return context$2$0.abrupt('return', (0, _utils.nok)('Xcode is NOT installed!'));

          case 12:
            context$2$0.t1 = xcodePath;

            if (!context$2$0.t1) {
              context$2$0.next = 17;
              break;
            }

            context$2$0.next = 16;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(xcodePath));

          case 16:
            context$2$0.t1 = context$2$0.sent;

          case 17:
            if (!context$2$0.t1) {
              context$2$0.next = 21;
              break;
            }

            context$2$0.t2 = (0, _utils.ok)('Xcode is installed at: ' + xcodePath);
            context$2$0.next = 22;
            break;

          case 21:
            context$2$0.t2 = (0, _utils.nok)('Xcode cannot be found at \'' + xcodePath + '\'!');

          case 22:
            return context$2$0.abrupt('return', context$2$0.t2);

          case 23:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 9]]);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', 'Manually install Xcode.');

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return XcodeCheck;
})(_doctor.DoctorCheck);

checks.push(new XcodeCheck());

// Check for Xcode Command Line Tools.

var XcodeCmdLineToolsCheck = (function (_DoctorCheck2) {
  _inherits(XcodeCmdLineToolsCheck, _DoctorCheck2);

  function XcodeCmdLineToolsCheck() {
    _classCallCheck(this, XcodeCmdLineToolsCheck);

    _get(Object.getPrototypeOf(XcodeCmdLineToolsCheck.prototype), 'constructor', this).call(this, { autofix: true });
  }

  _createClass(XcodeCmdLineToolsCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var errMess, pkgName, stdout;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            errMess = 'Xcode Command Line Tools are NOT installed!';
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap(_appiumSupport.system.macOsxVersion());

          case 3:
            context$2$0.t0 = context$2$0.sent;

            if (!(context$2$0.t0 === '10.8')) {
              context$2$0.next = 8;
              break;
            }

            context$2$0.t1 = 'com.apple.pkg.DeveloperToolsCLI';
            context$2$0.next = 9;
            break;

          case 8:
            context$2$0.t1 = 'com.apple.pkg.CLTools_Executables';

          case 9:
            pkgName = context$2$0.t1;
            stdout = undefined;
            context$2$0.prev = 11;
            context$2$0.next = 14;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('pkgutil', ['--pkg-info=' + pkgName]));

          case 14:
            stdout = context$2$0.sent.stdout;
            context$2$0.next = 21;
            break;

          case 17:
            context$2$0.prev = 17;
            context$2$0.t2 = context$2$0['catch'](11);

            _logger2['default'].debug(context$2$0.t2);
            return context$2$0.abrupt('return', (0, _utils.nok)(errMess));

          case 21:
            return context$2$0.abrupt('return', stdout.match(/install-time/) ? (0, _utils.ok)('Xcode Command Line Tools are installed.') : (0, _utils.nok)(errMess));

          case 22:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[11, 17]]);
    }
  }, {
    key: 'fix',
    value: function fix() {
      var yesno;
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _logger2['default'].info('The following command need be executed: xcode-select --install');
            context$2$0.next = 3;
            return _regeneratorRuntime.awrap((0, _prompt.fixIt)());

          case 3:
            yesno = context$2$0.sent;

            if (!(yesno === 'yes')) {
              context$2$0.next = 9;
              break;
            }

            context$2$0.next = 7;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('xcode-select', ['--install']));

          case 7:
            context$2$0.next = 11;
            break;

          case 9:
            _logger2['default'].info('Skipping you will need to install Xcode manually.');
            throw new _doctor.FixSkippedError();

          case 11:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return XcodeCmdLineToolsCheck;
})(_doctor.DoctorCheck);

checks.push(new XcodeCmdLineToolsCheck());

// Automatically run authorize iOS if requested
fixes.authorizeIosFix = function callee$0$0() {
  var yesno;
  return _regeneratorRuntime.async(function callee$0$0$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        _logger2['default'].info('The authorize iOS script need to be run.');
        context$1$0.next = 3;
        return _regeneratorRuntime.awrap((0, _prompt.fixIt)());

      case 3:
        yesno = context$1$0.sent;

        if (!(yesno === 'yes')) {
          context$1$0.next = 9;
          break;
        }

        context$1$0.next = 7;
        return _regeneratorRuntime.awrap((0, _utils.authorize)());

      case 7:
        context$1$0.next = 11;
        break;

      case 9:
        _logger2['default'].info('Skipping you will need to run the authorize iOS manually.');
        throw new _doctor.FixSkippedError();

      case 11:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
};

// Dev Tools Security

var DevToolsSecurityCheck = (function (_DoctorCheck3) {
  _inherits(DevToolsSecurityCheck, _DoctorCheck3);

  function DevToolsSecurityCheck() {
    _classCallCheck(this, DevToolsSecurityCheck);

    _get(Object.getPrototypeOf(DevToolsSecurityCheck.prototype), 'constructor', this).call(this, { autofix: true });
  }

  _createClass(DevToolsSecurityCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var errMess, stdout;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            errMess = 'DevToolsSecurity is NOT enabled!';
            stdout = undefined;
            context$2$0.prev = 2;
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('DevToolsSecurity', []));

          case 5:
            stdout = context$2$0.sent.stdout;
            context$2$0.next = 12;
            break;

          case 8:
            context$2$0.prev = 8;
            context$2$0.t0 = context$2$0['catch'](2);

            _logger2['default'].debug(context$2$0.t0);
            return context$2$0.abrupt('return', (0, _utils.nok)(errMess));

          case 12:
            return context$2$0.abrupt('return', stdout && stdout.match(/enabled/) ? (0, _utils.ok)('DevToolsSecurity is enabled.') : (0, _utils.nok)(errMess));

          case 13:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[2, 8]]);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(fixes.authorizeIosFix());

          case 2:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return DevToolsSecurityCheck;
})(_doctor.DoctorCheck);

checks.push(new DevToolsSecurityCheck());

// Authorization DB

var AuthorizationDbCheck = (function (_DoctorCheck4) {
  _inherits(AuthorizationDbCheck, _DoctorCheck4);

  function AuthorizationDbCheck() {
    _classCallCheck(this, AuthorizationDbCheck);

    _get(Object.getPrototypeOf(AuthorizationDbCheck.prototype), 'constructor', this).call(this, { autofix: true });
  }

  _createClass(AuthorizationDbCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var successMess, errMess, stdout, data, rg;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            successMess = 'The Authorization DB is set up properly.';
            errMess = 'The Authorization DB is NOT set up properly.';
            stdout = undefined;
            context$2$0.prev = 3;
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('security', ['authorizationdb', 'read', 'system.privilege.taskport']));

          case 6:
            stdout = context$2$0.sent.stdout;
            context$2$0.next = 32;
            break;

          case 9:
            context$2$0.prev = 9;
            context$2$0.t0 = context$2$0['catch'](3);
            context$2$0.next = 13;
            return _regeneratorRuntime.awrap(_appiumSupport.system.macOsxVersion());

          case 13:
            context$2$0.t1 = context$2$0.sent;

            if (!(context$2$0.t1 === '10.8')) {
              context$2$0.next = 30;
              break;
            }

            data = undefined;
            context$2$0.prev = 16;
            context$2$0.next = 19;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile('/etc/authorization', 'utf8'));

          case 19:
            data = context$2$0.sent;
            context$2$0.next = 26;
            break;

          case 22:
            context$2$0.prev = 22;
            context$2$0.t2 = context$2$0['catch'](16);

            _logger2['default'].debug(context$2$0.t2);
            return context$2$0.abrupt('return', (0, _utils.nok)(errMess));

          case 26:
            rg = /<key>system.privilege.taskport<\/key>\s*\n\s*<dict>\n\s*<key>allow-root<\/key>\n\s*(<true\/>)/;
            return context$2$0.abrupt('return', data && data.match(rg) ? (0, _utils.ok)(successMess) : (0, _utils.nok)(errMess));

          case 30:
            _logger2['default'].debug(context$2$0.t0);
            return context$2$0.abrupt('return', (0, _utils.nok)(errMess));

          case 32:
            return context$2$0.abrupt('return', stdout && (stdout.match(/is-developer/) || stdout.match(/allow/)) ? (0, _utils.ok)(successMess) : (0, _utils.nok)(errMess));

          case 33:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[3, 9], [16, 22]]);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(fixes.authorizeIosFix());

          case 2:
            return context$2$0.abrupt('return', context$2$0.sent);

          case 3:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return AuthorizationDbCheck;
})(_doctor.DoctorCheck);

checks.push(new AuthorizationDbCheck());

// Check for Carthage (for WDA)

var CarthageCheck = (function (_DoctorCheck5) {
  _inherits(CarthageCheck, _DoctorCheck5);

  function CarthageCheck() {
    _classCallCheck(this, CarthageCheck);

    _get(Object.getPrototypeOf(CarthageCheck.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(CarthageCheck, [{
    key: 'diagnose',
    value: function diagnose() {
      var carthagePath;
      return _regeneratorRuntime.async(function diagnose$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(_carthageDetector2['default'].detect());

          case 2:
            carthagePath = context$2$0.sent;
            return context$2$0.abrupt('return', carthagePath ? (0, _utils.ok)('Carthage was found at: ' + carthagePath) : (0, _utils.nok)('Carthage was NOT found!'));

          case 4:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }, {
    key: 'fix',
    value: function fix() {
      return _regeneratorRuntime.async(function fix$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            return context$2$0.abrupt('return', 'Please install Carthage. Visit https://github.com/Carthage' + '/Carthage#installing-carthage for more information.');

          case 1:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return CarthageCheck;
})(_doctor.DoctorCheck);

checks.push(new CarthageCheck());

checks.push(new _env2['default']('HOME'));

exports.fixes = fixes;
exports.XcodeCheck = XcodeCheck;
exports.XcodeCmdLineToolsCheck = XcodeCmdLineToolsCheck;
exports.DevToolsSecurityCheck = DevToolsSecurityCheck;
exports.AuthorizationDbCheck = AuthorizationDbCheck;
exports.CarthageCheck = CarthageCheck;
exports['default'] = checks;
