// appium-doctor/build/lib/node-detector.js
'use strict';

var _createClass = require('babel-runtime/helpers/create-class')['default'];

var _classCallCheck = require('babel-runtime/helpers/class-call-check')['default'];

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

var _getIterator = require('babel-runtime/core-js/get-iterator')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _appiumSupport = require('appium-support');

var _teen_process = require('teen_process');

var _logger = require('./logger');

var _logger2 = _interopRequireDefault(_logger);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var NODE_COMMON_PATHS = [process.env.NODE_BIN, '/usr/local/bin/node', '/opt/local/bin/node'];

// Look for node

var NodeDetector = (function () {
  function NodeDetector() {
    _classCallCheck(this, NodeDetector);
  }

  _createClass(NodeDetector, null, [{
    key: 'retrieveInCommonPlaces',
    value: function retrieveInCommonPlaces() {
      var _iteratorNormalCompletion, _didIteratorError, _iteratorError, _iterator, _step, p;

      return _regeneratorRuntime.async(function retrieveInCommonPlaces$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            _iteratorNormalCompletion = true;
            _didIteratorError = false;
            _iteratorError = undefined;
            context$2$0.prev = 3;
            _iterator = _getIterator(NODE_COMMON_PATHS);

          case 5:
            if (_iteratorNormalCompletion = (_step = _iterator.next()).done) {
              context$2$0.next = 18;
              break;
            }

            p = _step.value;
            context$2$0.t0 = p;

            if (!context$2$0.t0) {
              context$2$0.next = 12;
              break;
            }

            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(p));

          case 11:
            context$2$0.t0 = context$2$0.sent;

          case 12:
            if (!context$2$0.t0) {
              context$2$0.next = 15;
              break;
            }

            _logger2['default'].debug('Node binary found at common place: ' + p);
            return context$2$0.abrupt('return', p);

          case 15:
            _iteratorNormalCompletion = true;
            context$2$0.next = 5;
            break;

          case 18:
            context$2$0.next = 24;
            break;

          case 20:
            context$2$0.prev = 20;
            context$2$0.t1 = context$2$0['catch'](3);
            _didIteratorError = true;
            _iteratorError = context$2$0.t1;

          case 24:
            context$2$0.prev = 24;
            context$2$0.prev = 25;

            if (!_iteratorNormalCompletion && _iterator['return']) {
              _iterator['return']();
            }

          case 27:
            context$2$0.prev = 27;

            if (!_didIteratorError) {
              context$2$0.next = 30;
              break;
            }

            throw _iteratorError;

          case 30:
            return context$2$0.finish(27);

          case 31:
            return context$2$0.finish(24);

          case 32:
            _logger2['default'].debug('Node binary wasn\'t found at common places.');
            return context$2$0.abrupt('return', null);

          case 34:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[3, 20, 24, 32], [25,, 27, 31]]);
    }
  }, {
    key: 'retrieveUsingSystemCall',
    value: function retrieveUsingSystemCall() {
      var stdout, cmd, nodePath;
      return _regeneratorRuntime.async(function retrieveUsingSystemCall$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            stdout = undefined;
            cmd = 'which';

            if (_appiumSupport.system.isWindows()) {
              cmd = 'where';
            }

            context$2$0.prev = 3;
            context$2$0.next = 6;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)(cmd, ['node']));

          case 6:
            stdout = context$2$0.sent.stdout;
            context$2$0.next = 13;
            break;

          case 9:
            context$2$0.prev = 9;
            context$2$0.t0 = context$2$0['catch'](3);

            _logger2['default'].debug(context$2$0.t0);
            return context$2$0.abrupt('return', null);

          case 13:
            nodePath = stdout.replace(/[\n\r]/g, "");
            context$2$0.next = 16;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(nodePath));

          case 16:
            if (!context$2$0.sent) {
              context$2$0.next = 21;
              break;
            }

            _logger2['default'].debug('Node binary found using ' + cmd + ' command at: ' + nodePath);
            return context$2$0.abrupt('return', nodePath);

          case 21:
            _logger2['default'].debug('Node binary not found using the ' + cmd + ' command.');
            return context$2$0.abrupt('return', null);

          case 23:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[3, 9]]);
    }
  }, {
    key: 'retrieveUsingAppleScript',
    value: function retrieveUsingAppleScript() {
      var appScript, stdout, nodePath;
      return _regeneratorRuntime.async(function retrieveUsingAppleScript$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            if (_appiumSupport.system.isMac()) {
              context$2$0.next = 3;
              break;
            }

            _logger2['default'].debug('Not on Darwin, skipping Apple Script');
            return context$2$0.abrupt('return', null);

          case 3:
            appScript = ['try', '  set appiumIsRunning to false', '  tell application "System Events"', '    set appiumIsRunning to name of every process contains "Appium"', '  end tell', '  if appiumIsRunning then', '    tell application "Appium" to return node path', '  end if', 'end try', 'return "NULL"'].join("\n");
            stdout = undefined;
            context$2$0.prev = 5;
            context$2$0.next = 8;
            return _regeneratorRuntime.awrap((0, _teen_process.exec)('osascript', ['-e', appScript]));

          case 8:
            stdout = context$2$0.sent.stdout;
            context$2$0.next = 15;
            break;

          case 11:
            context$2$0.prev = 11;
            context$2$0.t0 = context$2$0['catch'](5);

            _logger2['default'].debug(context$2$0.t0);
            return context$2$0.abrupt('return', null);

          case 15:
            nodePath = stdout.replace("\n", "");
            context$2$0.next = 18;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(nodePath));

          case 18:
            if (!context$2$0.sent) {
              context$2$0.next = 23;
              break;
            }

            _logger2['default'].debug('Node binary found using AppleScript at: ' + nodePath);
            return context$2$0.abrupt('return', nodePath);

          case 23:
            _logger2['default'].debug('Node binary not found using AppleScript.');
            return context$2$0.abrupt('return', null);

          case 25:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[5, 11]]);
    }
  }, {
    key: 'retrieveUsingAppiumConfigFile',
    value: function retrieveUsingAppiumConfigFile() {
      var jsonobj, appiumConfigPath;
      return _regeneratorRuntime.async(function retrieveUsingAppiumConfigFile$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            jsonobj = undefined;
            context$2$0.prev = 1;
            appiumConfigPath = _path2['default'].resolve(__dirname, '..', '..', '.appiumconfig.json');
            context$2$0.next = 5;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(appiumConfigPath));

          case 5:
            if (!context$2$0.sent) {
              context$2$0.next = 11;
              break;
            }

            context$2$0.t0 = JSON;
            context$2$0.next = 9;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.readFile(appiumConfigPath, 'utf8'));

          case 9:
            context$2$0.t1 = context$2$0.sent;
            jsonobj = context$2$0.t0.parse.call(context$2$0.t0, context$2$0.t1);

          case 11:
            context$2$0.next = 17;
            break;

          case 13:
            context$2$0.prev = 13;
            context$2$0.t2 = context$2$0['catch'](1);

            _logger2['default'].debug(context$2$0.t2);
            return context$2$0.abrupt('return', null);

          case 17:
            context$2$0.t3 = jsonobj && jsonobj.node_bin;

            if (!context$2$0.t3) {
              context$2$0.next = 22;
              break;
            }

            context$2$0.next = 21;
            return _regeneratorRuntime.awrap(_appiumSupport.fs.exists(jsonobj.node_bin));

          case 21:
            context$2$0.t3 = context$2$0.sent;

          case 22:
            if (!context$2$0.t3) {
              context$2$0.next = 27;
              break;
            }

            _logger2['default'].debug('Node binary found using .appiumconfig.json at: ' + jsonobj.node_bin);
            return context$2$0.abrupt('return', jsonobj.node_bin);

          case 27:
            _logger2['default'].debug('Node binary not found in the .appiumconfig.json file.');
            return context$2$0.abrupt('return', null);

          case 29:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this, [[1, 13]]);
    }
  }, {
    key: 'detect',
    value: function detect() {
      var nodePath;
      return _regeneratorRuntime.async(function detect$(context$2$0) {
        while (1) switch (context$2$0.prev = context$2$0.next) {
          case 0:
            context$2$0.next = 2;
            return _regeneratorRuntime.awrap(NodeDetector.retrieveUsingSystemCall());

          case 2:
            context$2$0.t2 = context$2$0.sent;

            if (context$2$0.t2) {
              context$2$0.next = 7;
              break;
            }

            context$2$0.next = 6;
            return _regeneratorRuntime.awrap(NodeDetector.retrieveInCommonPlaces());

          case 6:
            context$2$0.t2 = context$2$0.sent;

          case 7:
            context$2$0.t1 = context$2$0.t2;

            if (context$2$0.t1) {
              context$2$0.next = 12;
              break;
            }

            context$2$0.next = 11;
            return _regeneratorRuntime.awrap(NodeDetector.retrieveUsingAppleScript());

          case 11:
            context$2$0.t1 = context$2$0.sent;

          case 12:
            context$2$0.t0 = context$2$0.t1;

            if (context$2$0.t0) {
              context$2$0.next = 17;
              break;
            }

            context$2$0.next = 16;
            return _regeneratorRuntime.awrap(NodeDetector.retrieveUsingAppiumConfigFile());

          case 16:
            context$2$0.t0 = context$2$0.sent;

          case 17:
            nodePath = context$2$0.t0;

            if (!nodePath) {
              context$2$0.next = 22;
              break;
            }

            return context$2$0.abrupt('return', nodePath);

          case 22:
            _logger2['default'].warn('The node binary could not be found.');
            return context$2$0.abrupt('return', null);

          case 24:
          case 'end':
            return context$2$0.stop();
        }
      }, null, this);
    }
  }]);

  return NodeDetector;
})();

exports['default'] = NodeDetector;
module.exports = exports['default'];
