// appium-doctor/build/lib/prompt.js
'use strict';

var _regeneratorRuntime = require('babel-runtime/regenerator')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _utils = require('./utils');

var persistentResponse = undefined;

var fixItQuestion = {
  type: 'list',
  name: 'confirmation',
  message: 'Fix it:',
  choices: ['yes', 'no', 'always', 'never'],
  filter: function filter(val) {
    return val.toLowerCase();
  }
};

function configure(opts) {
  if (opts.yes) {
    persistentResponse = 'yes';
  }
  if (opts.no) {
    persistentResponse = 'no';
  }
}

function clear() {
  persistentResponse = undefined;
}

function fixIt() {
  var resp;
  return _regeneratorRuntime.async(function fixIt$(context$1$0) {
    while (1) switch (context$1$0.prev = context$1$0.next) {
      case 0:
        if (!persistentResponse) {
          context$1$0.next = 2;
          break;
        }

        return context$1$0.abrupt('return', persistentResponse);

      case 2:
        context$1$0.next = 4;
        return _regeneratorRuntime.awrap(_utils.inquirer.prompt(fixItQuestion));

      case 4:
        resp = context$1$0.sent;

        persistentResponse = resp.confirmation === 'always' ? 'yes' : persistentResponse;
        persistentResponse = resp.confirmation === 'never' ? 'no' : persistentResponse;
        return context$1$0.abrupt('return', persistentResponse || resp.confirmation);

      case 8:
      case 'end':
        return context$1$0.stop();
    }
  }, null, this);
}

exports.configure = configure;
exports.fixIt = fixIt;
exports.clear = clear;
