// appium-doctor/build/lib/utils.js
'use strict';

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _bluebird = require('bluebird');

var _bluebird2 = _interopRequireDefault(_bluebird);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _inquirer2 = require('inquirer');

var _inquirer3 = _interopRequireDefault(_inquirer2);

var _libLogger = require('../lib/logger');

var _libLogger2 = _interopRequireDefault(_libLogger);

var _authorizeIos = require('authorize-ios');

var _authorizeIos2 = _interopRequireDefault(_authorizeIos);

var pkgRoot = process.env.NO_PRECOMPILE ? _path2['default'].resolve(__dirname, '..') : _path2['default'].resolve(__dirname, '..', '..');

var ok = function ok(message) {
  return { ok: true, message: message };
};
var nok = function nok(message) {
  return { ok: false, message: message };
};

var inquirer = {
  prompt: _bluebird2['default'].promisify(function (question, cb) {
    _inquirer3['default'].prompt(question, function (resp) {
      cb(null, resp);
    });
  })
};

function configureBinaryLog(opts) {
  var actualLog = _libLogger2['default'].unwrap().log;
  _libLogger2['default'].unwrap().log = function (level, prefix, msg) {
    var l = this.levels[level];
    if (l < this.levels[this.level]) return;
    actualLog(level, prefix, msg);
  };
  _libLogger2['default'].level = opts.debug ? 'debug' : 'info';
}

exports.pkgRoot = pkgRoot;
exports.ok = ok;
exports.nok = nok;
exports.inquirer = inquirer;
exports.configureBinaryLog = configureBinaryLog;
exports.authorize = _authorizeIos2['default'];
